server/sonar-server/src/main/java/org/sonar/server/permission/ws/PermissionsWsParametersBuilder.java

```diff
@@ -100,33 +120,90 @@ public static void createTemplateParameters(NewAction action) {
   }
 
   private static void createTemplateIdParameter(NewAction action) {
-    action.createParam(org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_TEMPLATE_ID)
+    action.createParam(PARAM_TEMPLATE_ID)
       .setDescription("Template id")
       .setExampleValue(Uuids.UUID_EXAMPLE_01);
   }
 
   private static void createTemplateNameParameter(NewAction action) {
-    action.createParam(org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_TEMPLATE_NAME)
+    action.createParam(PARAM_TEMPLATE_NAME)
       .setDescription("Template name")
       .setExampleValue("Default Permission Template for Projects");
   }
 
   public static void createTemplateProjectKeyPatternParameter(NewAction action) {
-    action.createParam(org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_PROJECT_KEY_PATTERN)
+    action.createParam(PARAM_PROJECT_KEY_PATTERN)
       .setDescription("Project key pattern. Must be a valid Java regular expression")
       .setExampleValue(".*\\.finance\\..*");
   }
 
   public static void createTemplateDescriptionParameter(NewAction action) {
-    action.createParam(org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_DESCRIPTION)
+    action.createParam(PARAM_DESCRIPTION)
       .setDescription("Description")
       .setExampleValue("Permissions for all projects related to the financial service");
   }
 
   public static void createIdParameter(NewAction action) {
-    action.createParam(org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_ID)
+    action.createParam(PARAM_ID)
       .setRequired(true)
       .setDescription("Id")
       .setExampleValue("af8cb8cc-1e78-4c4e-8c00-ee8e814009a5");
   }
+
+  public static NewParam createQualifierParameter(NewAction action, QualifierParameterContext context) {
+    return action.createParam(PARAM_QUALIFIER)
+      .setDescription("Project qualifier. Possible values are:" + buildRootQualifiersDescription(context))
+      .setPossibleValues(getRootQualifiers(context.getResourceTypes()));
+  }
+
+  private static Set<String> getRootQualifiers(ResourceTypes resourceTypes) {
+    return from(resourceTypes.getRoots())
+      .transform(RESOURCE_TYPE_TO_QUALIFIER)
+      .toSortedSet(natural());
+  }
+
+  private static String buildRootQualifiersDescription(QualifierParameterContext context) {
+    StringBuilder description = new StringBuilder();
+    description.append("<ul>");
+    String qualifierPattern = "<li>%s - %s</li>";
+    for (String qualifier : getRootQualifiers(context.getResourceTypes())) {
+      description.append(format(qualifierPattern, qualifier, i18n(context, qualifier)));
+    }
+    description.append("</ul>");
+
+    return description.toString();
+  }
+
+  private static String i18n(QualifierParameterContext context, String qualifier) {
```

Comment:
Method name `i18n` is not intuitive out of context. What about `qualifierLabel` ?